File: 2467_most_profitable_path_in_a_tree.py
from collections import defaultdict

class Solution:
    def mostProfitablePath(self, edges, bob, amount):
        g, m, mx = defaultdict(list), {}, float('-inf')

        for u, v in edges:
            g[u].append(v)
            g[v].append(u)

        def dfs_bob(n, t, v):
            v.add(n)
            m[n] = t
            if n == 0: return True
            for nb in g[n]:
                if nb not in v and dfs_bob(nb, t + 1, v):
                    return True
            del m[n]
            return False

        def dfs_alice(n, t, p, v):
            nonlocal mx
            v.add(n)
            p += amount[n] if n not in m or t < m[n] else amount[n] // 2 if t == m[n] else 0
            if len(g[n]) == 1 and n != 0:
                mx = max(mx, p)
            for nb in g[n]:
                if nb not in v:
                    dfs_alice(nb, t + 1, p, v)

        dfs_bob(bob, 0, set())
        dfs_alice(0, 0, 0, set())
        return mx

File: 2467_most_profitable_path_in_a_tree.java
import java.util.*;

class Solution {
    Map<Integer, List<Integer>> g = new HashMap<>();
    Map<Integer, Integer> bobMap = new HashMap<>();
    int maxProfit = Integer.MIN_VALUE;

    boolean dfsBob(int node, int time, boolean[] visited) {
        visited[node] = true;
        bobMap.put(node, time);
        if (node == 0) return true;
        for (int neighbor : g.get(node)) {
            if (!visited[neighbor] && dfsBob(neighbor, time + 1, visited)) {
                return true;
            }
        }
        bobMap.remove(node);
        return false;
    }

    void dfsAlice(int node, int time, int profit, boolean[] visited, int[] amount) {
        visited[node] = true;
        if (!bobMap.containsKey(node) || time < bobMap.get(node)) {
            profit += amount[node];
        } else if (time == bobMap.get(node)) {
            profit += amount[node] / 2;
        }
        if (g.get(node).size() == 1 && node != 0) {
            maxProfit = Math.max(maxProfit, profit);
        }
        for (int neighbor : g.get(node)) {
            if (!visited[neighbor]) {
                dfsAlice(neighbor, time + 1, profit, visited, amount);
            }
        }
    }

    public int mostProfitablePath(int[][] edges, int bob, int[] amount) {
        for (int[] e : edges) {
            g.computeIfAbsent(e[0], k -> new ArrayList<>()).add(e[1]);
            g.computeIfAbsent(e[1], k -> new ArrayList<>()).add(e[0]);
        }
        boolean[] visited = new boolean[amount.length];
        dfsBob(bob, 0, visited);
        Arrays.fill(visited, false);
        dfsAlice(0, 0, 0, visited, amount);
        return maxProfit;
    }
}
